### PR TITLE
Move Events to Rebus (Part 1)

### PR DESCRIPTION
**Pull request by barronpm**
**Changes**
Uses Rebus to handle eventing. Long-running tasks such as scans could possibly be moved to Rebus in the future as well. Needs further testing.

### CODE REVIEW

**Comment by crobibero** [comment]
IMO Emby.Notifications is basically useless, the notifications that can be received are just strings.
Plugins should listen to specific events, most can be deprecated and users can move to https://github.com/crobibero/jellyfin-plugin-webhook

**Comment by barronpm** [comment]
Currently, the only non-archived plugins that use the notifications are:

- jellyfin-plugin-email
- jellyfin-plugin-pushbullet
- jellyfin-plugin-slack

AFAIK these will all be replaced with @crobibero 's webhook plugin

**Review by crobibero — APPROVED** [review]

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 1]
```diff
@@ -206,30 +170,14 @@ private void OnSessionStarted(SessionInfo info)
...
-
-            EventHelper.QueueEventIfNotNull(
-                SessionStarted,
-                this,
-                new SessionEventArgs
-                {
-                    SessionInfo = info
-                },
-                _logger);
+            // TODO: fix
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:173]:
```suggestion
            // TODO: make the entire call async
```

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 2]
```diff
@@ -206,30 +170,14 @@ private void OnSessionStarted(SessionInfo info)
...
-                SessionEnded,
-                this,
-                new SessionEventArgs
-                {
-                    SessionInfo = info
-                },
-                _logger);
-
-            _eventManager.Publish(new SessionEndedEventArgs(info));
+            // TODO: fix
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:179]:
```suggestion
            // TODO: make the entire call async
```

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 3]
```diff
@@ -304,28 +252,17 @@ public void UpdateDeviceName(string sessionId, string deviceName)
 
             if ((activityDate - lastActivityDate).TotalSeconds > 10)
             {
-                SessionActivity?.Invoke(
-                    this,
-                    new SessionEventArgs
-                    {
-                        SessionInfo = session
-                    });
+                // TODO: fix
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:255]:
```suggestion
                // TODO: make the entire call async
```

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 4]
```diff
@@ -304,28 +252,17 @@ public void UpdateDeviceName(string sessionId, string deviceName)
...
+                _eventBus.Send(new SessionActivityEventArgs(session)).GetAwaiter().GetResult();
             }
 
             return session;
         }
 
         /// <inheritdoc />
-        public void OnSessionControllerConnected(SessionInfo info)
+        public async Task OnSessionControllerConnected(SessionInfo info)
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:263]:
```suggestion
        public Task OnSessionControllerConnected(SessionInfo info)
```

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 5]
```diff
@@ -304,28 +252,17 @@ public void UpdateDeviceName(string sessionId, string deviceName)
...
         {
-            EventHelper.QueueEventIfNotNull(
-                SessionControllerConnected,
-                this,
-                new SessionEventArgs
-                {
-                    SessionInfo = info
-                },
-                _logger);
+            await _eventBus.Send(new SessionControllerConnectedEventArgs(info)).ConfigureAwait(false);
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:265]:
```suggestion
            return _eventBus.Send(new SessionControllerConnectedEventArgs(info));
```

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 6]
```diff
@@ -1683,12 +1602,8 @@ public void ReportCapabilities(string sessionId, ClientCapabilities capabilities
 
             if (saveCapabilities)
             {
-                CapabilitiesChanged?.Invoke(
-                    this,
-                    new SessionEventArgs
-                    {
-                        SessionInfo = session
-                    });
+                // TODO: fix
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionManager.cs:1605]:
```suggestion
                // TODO: make the entire call async
```

File: Emby.Server.Implementations/Session/SessionWebSocketListener.cs [hunk 7]
```diff
@@ -100,7 +100,7 @@ public async Task ProcessWebSocketConnectedAsync(IWebSocketConnection connection
             var session = GetSession(connection.QueryString, connection.RemoteEndPoint.ToString());
             if (session != null)
             {
-                EnsureController(session, connection);
+                await EnsureController(session, connection);
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionWebSocketListener.cs:103]:
```suggestion
                await EnsureController(session, connection).ConfigureAwait(false);
```

File: Emby.Server.Implementations/Session/SessionWebSocketListener.cs [hunk 8]
```diff
@@ -126,15 +126,15 @@ private SessionInfo GetSession(IQueryCollection queryString, string remoteEndpoi
...
+        private async Task EnsureController(SessionInfo session, IWebSocketConnection connection)
         {
             var controllerInfo = session.EnsureController<WebSocketController>(
                 s => new WebSocketController(_loggerFactory.CreateLogger<WebSocketController>(), s, _sessionManager));
 
             var controller = (WebSocketController)controllerInfo.Item1;
             controller.AddWebSocket(connection);
 
-            _sessionManager.OnSessionControllerConnected(session);
+            await _sessionManager.OnSessionControllerConnected(session);
```
Comment (cvium) [inline Emby.Server.Implementations/Session/SessionWebSocketListener.cs:137]:
```suggestion
            await _sessionManager.OnSessionControllerConnected(session).ConfigureAwait(false);
```

File: Jellyfin.Server.Implementations/Users/UserManager.cs [hunk 9]
```diff
@@ -270,9 +266,9 @@ public Task ResetPassword(User user)
         }
 
         /// <inheritdoc/>
-        public void ResetEasyPassword(User user)
+        public async Task ResetEasyPassword(User user)
```
Comment (cvium) [inline Jellyfin.Server.Implementations/Users/UserManager.cs:269]:
```suggestion
        public Task ResetEasyPassword(User user)
```

File: Jellyfin.Server.Implementations/Users/UserManager.cs [hunk 10]
```diff
@@ -270,9 +266,9 @@ public Task ResetPassword(User user)
...
         {
-            ChangeEasyPassword(user, string.Empty, null);
+            await ChangeEasyPassword(user, string.Empty, null).ConfigureAwait(false);
```
Comment (cvium) [inline Jellyfin.Server.Implementations/Users/UserManager.cs:271]:
```suggestion
            return ChangeEasyPassword(user, string.Empty, null);
```

**Review by cvium — CHANGES_REQUESTED** [review]

**Review by crobibero — APPROVED** [review]

File: Emby.Server.Implementations/ScheduledTasks/TaskManager.cs [hunk 11]
```diff
@@ -19,8 +20,6 @@ public class TaskManager : ITaskManager
     {
         public event EventHandler<GenericEventArgs<IScheduledTaskWorker>> TaskExecuting;
```
Comment (cvium) [inline Emby.Server.Implementations/ScheduledTasks/TaskManager.cs:21]: What about this one?

Comment (barronpm) [reply Emby.Server.Implementations/ScheduledTasks/TaskManager.cs:21]: I only migrated some of the events in this PR, generally the ones that had already been at least partially moved to the custom event system. I'm working on followup PR(s) to migrate the rest of the events, to keep the size of each individual PR within reason.

**Review by crobibero — APPROVED** [review]
Tested all handlers, everything now passes without exception

**Review by crobibero — APPROVED** [review]

**Comment by barronpm** [comment]
Closing this, I plan on re-doing this PR with MediatR at some point

**Pull request closed without merging**